python/pyarrow/scalar.pxi

```diff
@@ -514,6 +514,32 @@ cdef class DurationScalar(Scalar):
             return datetime.timedelta(microseconds=sp.value // 1000)
 
 
+cdef class MonthDayNanoIntervalScalar(Scalar):
+    """
+    Concrete class for month, day, nanosecond scalars.
+    """
+
+    @property
+    def value(self):
+        """
+        Returns this value pyarrow.MonthDayNanoTuple.
+        """
+        cdef PyObject* val
+        val = GetResultValue(ARROW_TO_PYTHON.ToPrimitive(
+            (deref(self.wrapped.get()))))
+        return PyObject_to_object(val)
+
+    def as_py(self):
+        """
+        Return this value as a Pandas DateOffset instance if Pandas is present
+        otherwise as a named tuple containing months days and nanoseconds.
```

Comment:
I'm a bit concerned that this makes user code fragile: depending on whether Pandas is installed or not, you'll get an object with a different API.

Instead, perhaps always return a `pa.MonthDayNano`, but add a `MonthDayNano.to_pandas()` method to produce a Pandas DateOffset?

Reply:
I kind of agree, I modeled this after  [Timestamps](https://github.com/apache/arrow/blob/master/python/pyarrow/scalar.pxi#L382)  which also changes behavior.  Happy to simplify this though if that is desired (along with corresponding to_pylist()) logic.

Reply:
Hmm, it's a bit unfortunate for timestamps, but the decision there is understandable: returning a `datetime.datetime` loses precision. Also, `pandas.Timestamp` is probably API-compatible with `datetime.datetime`.

You don't lose any precision by returning a `MonthDayNano`.

Reply:
I'm going to agree with @pitrou .  With pyarrow we even have cases where users aren't even really aware they are using pyarrow.  I could see a circumstance where they install their application on some new machine or environment and suddenly start getting errors.  Maybe someday we can add a `use_pandas_types` option to `as_py`, `to_pylist`, and `to_pydict`.  That way we can error if the expected library is not available instead of silently changing the type.

Reply:
Going to change.  This actually simplifies the code a lot.

Reply:
Yes, I also agree returning a consistent type here is best (for Timestamp there is the complication of datetime.datetime not supporting nanoseconds ..). As Weston mentions, we can see later if we want to add some option to get a more pandas native object.